### PR TITLE
Example and utility for partial failure

### PR DESCRIPTION
**Pull request by Raibaz**

### CODE REVIEW

File: examples/BasicOperations/HandlePartialFailure.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+use GetOpt\GetOpt;
+use Google\Ads\GoogleAds\Examples\Utils\ArgumentNames;
+use Google\Ads\GoogleAds\Examples\Utils\ArgumentParser;
+use Google\Ads\GoogleAds\Lib\V1\GoogleAdsClient;
+use Google\Ads\GoogleAds\Lib\V1\GoogleAdsClientBuilder;
+use Google\Ads\GoogleAds\Lib\V1\GoogleAdsException;
+use Google\Ads\GoogleAds\Lib\OAuth2TokenBuilder;
+use Google\Ads\GoogleAds\Util\V1\ResourceNames;
+use Google\Ads\GoogleAds\V1\Errors\GoogleAdsError;
+//use Google\Ads\GoogleAds\V1\Services\GoogleAdsRow;
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: Remove

File: examples/BasicOperations/HandlePartialFailure.php [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * Runs the example.
+     *
+     * @param GoogleAdsClient $googleAdsClient the Google Ads API client
+     * @param int $customerId the client customer ID without hyphens
+     * @param int $campaignId a campaign ID
+     */
+    public static function runExample(GoogleAdsClient $googleAdsClient, $customerId, $campaignId)
+    {
+    
+        $temp_unused_var = new \Google\Ads\GoogleAds\V1\Errors\GoogleAdsFailure();
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: Use camelCase + do import of GoogleAdsFailure

File: examples/BasicOperations/HandlePartialFailure.php [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+        $campaignResourceName =
+            new StringValue(['value' => ResourceNames::forCampaign($customerId, $campaignId)]);
+
+        // This AdGroup should be created successfully - assuming the campaign in the params
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: Nit: This "ad group"

File: examples/BasicOperations/HandlePartialFailure.php [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            new StringValue(['value' => ResourceNames::forCampaign($customerId, $campaignId)]);
+
+        // This AdGroup should be created successfully - assuming the campaign in the params 
+        // exists.
+        $adGroup1 = new AdGroup([
+            'name' => new StringValue(['value' => 'Valid AdGroup #' . uniqid()]),
+            'campaign' => $campaignResourceName            
+        ]);
+
+        // This AdGroup will always fail - campaign ID 0 in resource names is never valid.
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: campaign ID 0 in "the resource name" is never valid.

File: examples/BasicOperations/HandlePartialFailure.php (hunk 3 above)
Comment (fiboknacky) [reply examples/BasicOperations/HandlePartialFailure.php]: Please fix all below occurrences too.

File: examples/BasicOperations/HandlePartialFailure.php [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        {
+            printf(
+                "All operations completed successfully. No partial failures to show.%s",
+                PHP_EOL
+            );
+        }
+
+        // Finds the failed operations by looping through the results.
+        $operationIndex = 0;
+        foreach($response->getResults() as $result)
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: Fix all PSR-2 problems here.
Always run phpcs first please.

File: examples/BasicOperations/HandlePartialFailure.php [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+        // Issues the mutate request, setting partialFailure=true.
+        $adGroupServiceClient = $googleAdsClient->getAdGroupServiceClient();
+        $response = $adGroupServiceClient->mutateAdGroups(
+            $customerId,
+            $operations,
+            ['partialFailure' => true]
+        );
+
+        if(!is_null($response->getPartialFailureError()))
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: This clause is not PSR-2 compliant.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    /**
+     * Checks if a result in a mutate response is a partial failure.
+     *
+     * @param \Google\Protobuf\Internal\Message $message
+     * @return bool
+     */
+    public static function isPartialFailure(Message $message)
+    {
+        return $message->byteSize() == 0;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Use ===

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    }
+ }
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Add another blank line pls.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 9]
```diff
@@ -0,0 +1,164 @@
+<?php
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Path is wrong. Please check with the existing utils.

File: examples/BasicOperations/HandlePartialFailure.php (hunk 2 above)
Comment (fiboknacky) [reply examples/BasicOperations/HandlePartialFailure.php]: And this shouldn't be here.
It should be at least very close to when we call mutateAdGroups and you need to explain why we need this.

It'd be better to make your ErrorUtils instantiable and include this line in the constructor instead.
Users wouldn't need to know about this detail.
You can provide this as a constructor parameter to help dependency injection.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php (hunk 9 above)
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Please add test too.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Ads\GoogleAds\V0\Util;
+
+use \Google\Protobuf\Any;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Prefix \ is not necessary.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+     *         }
+     *       }
+     *     }
+     *   </code>
+     * </pre>
+     *
+     * A single GoogleAdsError instance would be returned for operation index 1 and 2, and an empty
+     * list otherwise.
+     *
+     * @param int operationIndex the index of the operation, starting from 0.
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Nit: we don't end with a period for @param and @return

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+     * @param \Google\Rpc\Status partialFailureStatus a partialFailure status, with the detail list
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Include Status and don't use the full path here.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+     *     }
+     *   </code>
+     * </pre>
+     *
+     * A single GoogleAdsError instance would be returned for operation index 1 and 2, and an empty
+     * list otherwise.
+     *
+     * @param int operationIndex the index of the operation, starting from 0.
+     * @param \Google\Rpc\Status partialFailureStatus a partialFailure status, with the detail list 
+     *  containing GoogleAdsFailure instances.
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Need to indent for 4 spaces for hanging indentation.

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        $result = [];
+        foreach($failure->getErrors() as $error)
+        {
+            $pathElements = $error->getLocation()->getFieldPathElements();
+            if(count($pathElements) > 0)
+            {
+                $element = $pathElements[0];
+                $fieldName = $element->getFieldName();
+                $index = $element->getIndex();
+                if($fieldName == "operations" && $index->getValue() == $operationIndex)
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Use ===

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            $errors = self::getGoogleAdsErrorsFromFailure($operationIndex, $failure);
+            $result = array_merge($result, $errors);
+        }
+        return $result;
+    }
+
+    /**
+     * Return a list of GoogleAdsError instances for a given operation index.
+     *
+     * @see #getGoogleAdsErrorsFromStatus
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: @see ErrorUtils::getGoogleAdsErrorsFromStatus

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+     * list otherwise.
+     *
+     * @param int operationIndex the index of the operation, starting from 0.
+     * @param \Google\Rpc\Status partialFailureStatus a partialFailure status, with the detail list 
+     *  containing GoogleAdsFailure instances.
+     * @return Google\Ads\GoogleAds\V1\Errors\GoogleAdsError[] an array containing the 
+     *  GoogleAdsError instances for a given operation index.
+     */
+    public static function getGoogleAdsErrorsFromStatus(
+        int $operationIndex,
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Is this a valid PHP 7.1 syntax?

**Review by fiboknacky — COMMENTED** [review]
- Check your code styles in all places. It should comply with PSR-2. Always run phpcs before exporting PRs.
- Please apply the same comment to v1 util file too.
- Name of the utils would need a change. We stick to the Guava naming style, e.g., ResourceNames.

File: examples/BasicOperations/HandlePartialFailure.php (hunk 2 above)
Comment (Raibaz) [reply examples/BasicOperations/HandlePartialFailure.php]: I don't like much the idea of having the ErrorUtils class instantiable, but I agree that that initialization should not be there. I made an `ErrorUtils::initialize()` function that does this as close as possible to the mutate call but still allows to keep all functions static, as the ErrorUtils class does not in fact have any state.

WDYT?

File: src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php (hunk 16 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/V0/Util/ErrorUtils.php]: Yep scalar type hinting has been introduced in PHP 7: http://php.net/manual/en/migration70.new-features.php

File: examples/BasicOperations/HandlePartialFailure.php (hunk 2 above)
Comment (fiboknacky) [reply examples/BasicOperations/HandlePartialFailure.php]: Could you elaborate why you don't like it? Actually, it's used in that way in many of our PHP SOAP library utilities.
Being non-static actually helps when you write test too, as you don't need to worry about inadvertently touch the same parts inadvertently.

BTW, having object instantiation in the initialize() method smells like that we actually need a constructor too.

Comment (Raibaz) [reply examples/BasicOperations/HandlePartialFailure.php]: I don't like it because, although it contains an object instantiation, the ErrorUtils does not really have a state itself, so all the methods already in there can be used statically (and, in fact, they are).

The reason why we need to instantiate a GoogleAdsFailure is because of a quirk with the descriptor pool, not because the ErrorUtils class actually uses it or it keeps track of a GoogleAdsFailure object as its state.

As an alternative approach, what if we did the GoogleAdsFailure instantiation inside the `getGoogleAdsFailureFromAny` function in ErrorUtils?

It's the closest place to where it's actually needed and it would be transparent for clients of the ErrorUtils class, without needing to instantiate it or to call `initialize()`.

Comment (fiboknacky) [reply examples/BasicOperations/HandlePartialFailure.php]: > I don't like it because, although it contains an object instantiation, the ErrorUtils does not really have a state itself, so all the methods already in there can be used statically (and, in fact, they are).

> The reason why we need to instantiate a GoogleAdsFailure is because of a quirk with the descriptor pool, not because the ErrorUtils class actually uses it or it keeps track of a GoogleAdsFailure object as its state.

That's a fair point. I'm fine with that too.

Note that instantiation of GoogleAdsFailure needs to be done only before calling the method that uses partial failure.
So, getGoogleAdsFailureFromAny() cannot be used in this case.
initialize() seems the closest we can get.

Regarding the class structure, can you follow the style I said before?
Other examples than ResourceNames would be [CsvFiles](https://github.com/googleads/googleads-php-lib/blob/master/src/Google/AdsApi/AdManager/Util/v201902/CsvFiles.php) and [AdManagerDates](https://github.com/googleads/googleads-php-lib/blob/master/src/Google/AdsApi/AdManager/Util/v201902/AdManagerDates.php).

After looking at your methods, I think we should divide them into groups:

- PartialFailures contains initialize() (to instantiate GoogleAdsFailure)
- GoogleAdsError contains fromStatus() and fromFailure()
- GoogleAdsFailure contains fromAny() and fromStatus()

Make sure you add tests for them too.

Comment (Raibaz) [reply examples/BasicOperations/HandlePartialFailure.php]: Right, I split the ErrorUtils class like you mentioned ad added test coverage.

File: examples/BasicOperations/HandlePartialFailure.php [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+        $response = $adGroupServiceClient->mutateAdGroups(
+            $customerId,
+            $operations,
+            ['partialFailure' => true]
+        );
+
+        if (!is_null($response->getPartialFailureError())) {
+            printf("Partial failures occurred. Details will be shown below.%s", PHP_EOL);
+        } else {
+            printf(
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: This command has over 100 chars?

File: examples/BasicOperations/HandlePartialFailure.php [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+                "All operations completed successfully. No partial failures to show.%s",
+                PHP_EOL
+            );
+        }
+
+        // Finds the failed operations by looping through the results.
+        $operationIndex = 0;
+        foreach ($response->getResults() as $result) {
+            if (PartialFailures::isPartialFailure($result)) {
+                $errors = GoogleAdsErrors::fromStatus(
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: Can this fit in one line?

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Ads\GoogleAds\V0\Util;
+
+use \Google\Rpc\Status;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: Preceding \ wouldn't be needed.

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 20]
```diff
@@ -0,0 +1,118 @@
+<?php
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: You still put the file in the wrong directory.
It must be GoogleAds/Util/V0.

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+     *     }
+     *   </code>
+     * </pre>
+     *
+     * A single GoogleAdsError instance would be returned for operation index 1 and 2, and an empty
+     * list otherwise.
+     *
+     * @param int operationIndex the index of the operation, starting from 0.
+     * @param Status partialFailureStatus a partialFailure status, with the detail list
+     *    containing GoogleAdsFailure instances.
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: Again, @param and @return don't end with a period.

Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: And you need to start at hanging indented line at char 5, not char 4

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+     * @return GoogleAdsError[] an array containing the
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: Do you really need to indent this line?
Or at least, we should indent at later positions. This is a bit too early.

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+        foreach ($partialFailureStatus->getDetails() as $detail) {
+            $failure = GoogleAdsFailures::fromAny($detail);
+            $errors = self::fromFailure($operationIndex, $failure);
+            $result = array_merge($result, $errors);
+        }
+        return $result;
+    }
+
+    /**
+     * Return a list of GoogleAdsError instances for a given operation index.
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: "Returns"

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+    ) {
+        $result = [];
+        foreach ($failure->getErrors() as $error) {
+            $pathElements = $error->getLocation()->getFieldPathElements();
+            if (count($pathElements) > 0) {
+                /** @var FieldPathElement $element */
+                $element = $pathElements[0];
+                $fieldName = $element->getFieldName();
+                $index = $element->getIndex();
+                if ($fieldName === "operations" && $index->getValue() == $operationIndex) {
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: Use === everywhere when applicable.

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailures.php [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Ads\GoogleAds\V0\Util;
+
+use \Google\Protobuf\Any;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailures.php]: No \

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailures.php [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+use \Google\Rpc\Status;
+use Google\Ads\GoogleAds\V0\Errors\GoogleAdsFailure;
+
+final class GoogleAdsFailures
+{
+    /**
+     * Unpacks a single GoogleAdsFailure from an Any instance.
+     *
+     * @param Any An Any instance to unpack
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailures.php]: Don't need to capitalize the "An"

File: src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Ads\GoogleAds\V0\Util;
+
+use \Google\Protobuf\Internal\Message;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php]: No \

File: src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Checks if a result in a mutate response is a partial failure.
+     *
+     * @param Message $message
+     * @return bool
+     */
+    public static function isPartialFailure(Message $message)
+    {
+        return $message->byteSize() == 0;
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php]: I was told by the protobuf lib owner that it's not a good idea to use byteSize(), as the c extension doesn't have it.
He suggested to use serialize() and then check for its size instead.

File: tests/Google/Ads/GoogleAds/V1/Util/PartialFailuresTest.php [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+ * Unit tests for `PartialFailures`.
+ *
+ * @see PartialFailures
+ * @small
+ */
+class PartialFailuresTest extends TestCase
+{
+    public function testIsPartialFailure()
+    {
+        $messageMock = $this
```
Comment (fiboknacky) [inline tests/Google/Ads/GoogleAds/V1/Util/PartialFailuresTest.php]: Instead of mocking many thing, can you research to see if we can just instantiate a Message?
We should avoid doing "over-mocking", when possible.

**Review by fiboknacky — CHANGES_REQUESTED** [review]

File: src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php (hunk 28 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/V0/Util/PartialFailures.php]: Done.

File: src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php (hunk 20 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrors.php]: Sorry, I was mistaken, I just realized that this folder was the one for generated code. Fixed it.

File: tests/Google/Ads/GoogleAds/V1/Util/PartialFailuresTest.php (hunk 29 above)
Comment (Raibaz) [reply tests/Google/Ads/GoogleAds/V1/Util/PartialFailuresTest.php]: Agreed, I replaced mocks with actual object instances in most cases. I left the mocks where creating an object with the expected behavior was not immediate, such as driving `any->unpack` or message serialization.

File: examples/BasicOperations/HandlePartialFailure.php (hunk 18 above)
Comment (Raibaz) [reply examples/BasicOperations/HandlePartialFailure.php]: Nope :(

File: .gitignore [hunk 30]
```diff
@@ -1,3 +1,4 @@
 .idea
```
Comment (fiboknacky) [inline .gitignore]: This file shouldn't be part of this PR.

File: tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailuresTest.php [hunk 31]
```diff
@@ -15,70 +15,46 @@
...
-        $emptyStatusMock = $this
-            ->getMockBuilder(Status::class)
-            ->disableOriginalConstructor()
-            ->getMock();
-        $emptyStatusMock
-            ->method("getDetails")
-            ->willReturn([]);
-
-        $this->assertCount(0, GoogleAdsFailures::fromStatus($emptyStatusMock));
+        $emptyStatus = new Status();
```
Comment (fiboknacky) [inline tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailuresTest.php]: Nit: this assignment wouldn't be necessary.

File: src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+use Google\Ads\GoogleAds\V0\Errors\GoogleAdsFailure;
+
+final class PartialFailures
+{
+    /**
+     * Initializes a GoogleAdsFailure to make sure that the metadata pool knows about it.
+     */
+    public static function initialize()
+    {
+        $tempUnusedVar = new GoogleAdsFailure();
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php]: Heads-up: This wouldn't be needed once #41 is merged.

**Review by fiboknacky — CHANGES_REQUESTED** [review]
The tests look much much better and easier to read. Thanks for doing this. :)

File: src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /**
+     * Unpacks a single GoogleAdsFailure from an Any instance.
+     *
+     * @param Any an Any instance to unpack
+     * @return GoogleAdsFailure
+     */
+    public static function fromAny(Any $any)
+    {
+        return $any->unpack();
```
Comment (michaelbausor) [inline src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php]: A thought here - do you want to check that the unpacked message is actually the expected GoogleAdsFailure class, and if not, to either return null or throw an exception? Right now you could call this method on an Any object containing anything, and it will succeed and give you back that object.

File: src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php (hunk 32 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php]: Cool, then I think it's best to merge #41 first then rebase this on top of it and remove this function and its usage in the example.

WDYT?

File: src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php (hunk 33 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php]: Good point, especially since the clients are going to assume that the returned object is a GoogleAdsFailure. 

I think in this case throwing an exception would be the best solution, as unpacking a message that is not a GoogleAdsFailure here is really a wrong usage of this function. 

However, I don't have a strong opinion on what exception we should throw here - InvalidArgumentException? A custom exception? @fiboknacky, WDYT?

File: tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailuresTest.php (hunk 31 above)
Comment (Raibaz) [reply tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsFailuresTest.php]: Done.

File: .gitignore (hunk 30 above)
Comment (Raibaz) [reply .gitignore]: Done, sending a separate PR for this.

File: src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php (hunk 32 above)
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php]: Yep. I just merged.

File: src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php (hunk 33 above)
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php]: InvalidArgumentException SGTM.

File: src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php (hunk 32 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/Util/V0/PartialFailures.php]: Ok, removed this, thanks!

File: src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php (hunk 33 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/Util/V1/GoogleAdsFailures.php]: Done.

File: composer.json [hunk 34]
```diff
@@ -13,7 +13,7 @@
     "monolog/monolog": "^1.23.0"
   },
   "require-dev": {
-    "phpunit/phpunit": "^4.8.35 || ^5.7",
+    "phpunit/phpunit": "^7.5",
```
Comment (fiboknacky) [inline composer.json]: Probably need a rebase?

File: examples/BasicOperations/HandlePartialFailure.php [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+
+        $adGroupOperation2 = new AdGroupOperation();
+        $adGroupOperation2->setCreate($adGroup2);
+        $operations[] = $adGroupOperation2;
+
+        $adGroupOperation3 = new AdGroupOperation();
+        $adGroupOperation3->setCreate($adGroup3);
+        $operations[] = $adGroupOperation3;
+        
+        // Issues the mutate request, setting partialFailure=true.
```
Comment (fiboknacky) [inline examples/BasicOperations/HandlePartialFailure.php]: "Issues a mutate request by enabling partial failure mode."

File: src/Google/Ads/GoogleAds/Lib/V0/LogMessageFormatter.php [hunk 36]
```diff
@@ -114,6 +116,14 @@ public function formatDetail(
         );
 
         if ($status->code === 0) {
+            // Initialize unused GoogleAdsFailure object when there are partial failures returned.
```
Comment (fiboknacky) [inline src/Google/Ads/GoogleAds/Lib/V0/LogMessageFormatter.php]: Probably you need a rebase?

File: tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrorsTest.php [hunk 37]
```diff
@@ -0,0 +1,114 @@
+<?php
```
Comment (fiboknacky) [inline tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrorsTest.php]: Path still wrong.

Comment (fiboknacky) [reply tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrorsTest.php]: Please fix for all files of v0 and v1.

**Review by fiboknacky — CHANGES_REQUESTED** [review]

File: tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrorsTest.php (hunk 37 above)
Comment (Raibaz) [reply tests/Google/Ads/GoogleAds/V0/Util/GoogleAdsErrorsTest.php]: Done.

File: src/Google/Ads/GoogleAds/Lib/V0/LogMessageFormatter.php (hunk 36 above)
Comment (Raibaz) [reply src/Google/Ads/GoogleAds/Lib/V0/LogMessageFormatter.php]: Rebased.

File: composer.json (hunk 34 above)
Comment (Raibaz) [reply composer.json]: Rebased.

File: examples/BasicOperations/HandlePartialFailure.php (hunk 35 above)
Comment (fiboknacky) [reply examples/BasicOperations/HandlePartialFailure.php]: Not done.

**Review by fiboknacky — APPROVED** [review]
Please fix one small unresolved nit.
